konfetti/src/main/java/nl/dionsegijn/konfetti/models/Shape.kt

```diff
@@ -1,12 +1,93 @@
 package nl.dionsegijn.konfetti.models
 
-/**
- * Created by dionsegijn on 3/26/17.
- * Available shapes
- * RECT (rectangle)
- * Circle
- */
-enum class Shape {
-    RECT(),
-    CIRCLE()
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.PorterDuff
+import android.graphics.RectF
+import android.graphics.drawable.Drawable
+
+interface Shape {
+    /**
+     * Draw a shape to `canvas`. Implementations are expected to draw within a square of size
+     * `size` and must vertically/horizontally center their asset if it does not have an equal width
+     * and height.
+     */
+    fun draw(canvas: Canvas, paint: Paint, size: Float)
+
+    companion object {
+        // Maintain binary and backwards compatibility with previous enum API.
+        @JvmField
+        @Deprecated("Use Square class, instead.", replaceWith = ReplaceWith("Shape.Square"))
+        val RECT = Square
+
+        @JvmField
+        @Deprecated("Use Circle class, instead.", replaceWith = ReplaceWith("Shape.Circle"))
+        val CIRCLE = Circle
+    }
+
+    object Square : Shape {
+        override fun draw(
+            canvas: Canvas,
+            paint: Paint,
+            size: Float
+        ) {
+            canvas.drawRect(0f, 0f, size, size, paint)
+        }
+    }
+
+    class Rectangle(
+        /** The ratio of height to width. Must be within range [0, 1] */
+        private val sizeRatio: Float
+    ) : Shape {
+        init {
+            require(sizeRatio in 0f..1f)
+        }
+
+        override fun draw(canvas: Canvas, paint: Paint, size: Float) {
+            val height = size * sizeRatio
+            val top = (size - height) / 2f
+            canvas.drawRect(0f, top, size, top + height, paint)
+        }
+    }
+
+    object Circle : Shape {
+        private val rect = RectF()
+
+        override fun draw(
+            canvas: Canvas,
+            paint: Paint,
+            size: Float
+        ) {
+            rect.set(0f, 0f, size, size)
+            canvas.drawOval(rect, paint)
+        }
+    }
+
+    class DrawableShape(
+        drawable: Drawable,
+        /** Set to `false` to opt out of tinting. */
```

Comment:
Since I didn't understand it the first time, maybe changing this to something like:

```suggestion
        /** Set to false to opt out tinting the drawable and keep the original colours */
```

Reply:
Changed with slightly different wording.